domain/keymanager/state/state_test.go

```diff
@@ -0,0 +1,297 @@
+	c.Check(len(keys), gc.Equals, 0)
+}
+
+// TestDeletePublicKeysForNonExistentUser is asserting that if we try and
+// delete public keys for a user that doesn't exist we get back no error.
+func (s *stateSuite) TestDeletePublicKeysForNonExistentUser(c *gc.C) {
+	userId := usertesting.GenUserUUID(c)
+	state := NewState(s.TxnRunnerFactory())
+	err := state.DeletePublicKeysForUser(context.Background(), userId, []string{"comment"})
+	c.Check(err, jc.ErrorIsNil)
+}
+
+// TestDeletePublicKeysForComment is testing that we can remove a users public
+// keys via the comment string.
+func (s *stateSuite) TestDeletePublicKeysForComment(c *gc.C) {
+	state := NewState(s.TxnRunnerFactory())
+	keysToAdd := generatePublicKeys(c, testingPublicKeys)
+
+	err := state.AddPublicKeysForUser(context.Background(), s.userId, keysToAdd)
+	c.Check(err, jc.ErrorIsNil)
+
+	err = state.DeletePublicKeysForUser(context.Background(), s.userId, []string{
+		keysToAdd[0].Comment,
+	})
+	c.Assert(err, jc.ErrorIsNil)
+
+	keys, err := state.GetPublicKeysForUser(context.Background(), s.userId)
+	c.Assert(err, jc.ErrorIsNil)
+	fmt.Println(keys)
+	slices.Sort(keys)
+	slices.Sort(testingPublicKeys)
+	c.Check(testingPublicKeys[1:], jc.DeepEquals, keys)
+}
+
+// TestDeletePublicKeysForComment is testing that we can remove a users public
+// keys via the fingerprint.
+func (s *stateSuite) TestDeletePublicKeysForFingerprint(c *gc.C) {
+	state := NewState(s.TxnRunnerFactory())
+	keysToAdd := generatePublicKeys(c, testingPublicKeys)
+
+	err := state.AddPublicKeysForUser(context.Background(), s.userId, keysToAdd)
+	c.Check(err, jc.ErrorIsNil)
+
+	err = state.DeletePublicKeysForUser(context.Background(), s.userId, []string{
+		keysToAdd[0].Fingerprint,
+	})
+	c.Assert(err, jc.ErrorIsNil)
+
+	keys, err := state.GetPublicKeysForUser(context.Background(), s.userId)
+	c.Assert(err, jc.ErrorIsNil)
+	fmt.Println(keys)
+	slices.Sort(keys)
+	slices.Sort(testingPublicKeys)
+	c.Check(testingPublicKeys[1:], jc.DeepEquals, keys)
+}
+
+// TestDeletePublicKeysForComment is testing that we can remove a users public
+// keys via the keys data.
+func (s *stateSuite) TestDeletePublicKeysForKeyData(c *gc.C) {
+	state := NewState(s.TxnRunnerFactory())
+	keysToAdd := generatePublicKeys(c, testingPublicKeys)
+
+	err := state.AddPublicKeysForUser(context.Background(), s.userId, keysToAdd)
+	c.Check(err, jc.ErrorIsNil)
+
+	err = state.DeletePublicKeysForUser(context.Background(), s.userId, []string{
+		keysToAdd[0].Key,
+	})
+	c.Assert(err, jc.ErrorIsNil)
+
+	keys, err := state.GetPublicKeysForUser(context.Background(), s.userId)
+	c.Assert(err, jc.ErrorIsNil)
+	fmt.Println(keys)
+	slices.Sort(keys)
+	slices.Sort(testingPublicKeys)
+	c.Check(testingPublicKeys[1:], jc.DeepEquals, keys)
+}
+
+// TestDeletePublicKeysForCombination is asserting that we can remove a users
+// public keys via a combination of fingerprint and comment.
+func (s *stateSuite) TestDeletePublicKeysForCombination(c *gc.C) {
```

Comment:
Can we add a test if you remove the same key with both comment and fingerprint, to lock in the behaviour.

Lastly can we add tons of keys and remove just one.

Reply:
Happy to do the second test. Are we go doing key gen in the tests knowing how expensive it will be? Other option is to put junk data in but that hurts my head as it isn't in the spirit of the state.